File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/ChartDataDrillingDemo/ChartDataDrillingDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ChartDataDrillingDemoMain.h"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDataStorage"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridChartView"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBChartView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxGridCardView"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma resource "*.dfm"
TfrmMain *frmMain;
//---------------------------------------------------------------------------
__fastcall TfrmMain::TfrmMain(TComponent* Owner)
		: TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::FormCreate(TObject *Sender)
{
  String APath = ExtractFilePath(Application->ExeName) + "..\\..\\Data\\";
  tblCustomers->LoadFromFile(APath + "Customers.xml");
  tblProducts->LoadFromFile(APath + "Products.xml");
  tblOrders->LoadFromFile(APath + "Orders.xml");
};
//---------------------------------------------------------------------------
TcxGridDBColumn* TfrmMain::GetColumnByChartItem(TcxGridChartItem* AChartItem)
{
  TcxGridDBChartItemDataBinding* ADataBinding = (TcxGridDBChartItemDataBinding*)AChartItem->DataBinding;
  return TableView->GetColumnByFieldName(ADataBinding->FieldName);
}
//---------------------------------------------------------------------------

void TfrmMain::ShowTableActiveGroup()
{
  int ADataGroupIndex;

  // expand group rows so that the group with the active chart data is made visible
  ActiveDataGroups.Length = ChartView->ActiveDataLevel;
  TableView->BeginUpdate();
  try {
    TableView->DataController->Groups->FullCollapse();
    ADataGroupIndex = -1;
    for (int i = 0; i < ChartView->ActiveDataLevel; i++) {
      ADataGroupIndex = TableView->DataController->Groups->GetDataGroupIndexByGroupValue(
        ADataGroupIndex, ChartView->VisibleDataGroups[i]->ActiveValue);
      ActiveDataGroups[i] = ADataGroupIndex;

      int AGroupRowIndex = TableView->DataController->DataControllerInfo->DataGroups->Items[ADataGroupIndex]->RowIndex;
      TableView->DataController->Groups->ChangeExpanding(AGroupRowIndex, true, false);
      TableView->Controller->FocusedRowIndex = AGroupRowIndex;
    }
  }
  __finally {
    TableView->EndUpdate();
  }
  // select rows with the data used for the active chart
  TableView->BeginUpdate();
  try {
    if (ADataGroupIndex == -1)
      TableView->Controller->SelectAll();
    else {
      TableView->Controller->ClearSelection();
      for (int i = TableView->Controller->FocusedRowIndex + 1; i < TableView->ViewData->RowCount; i++) {
        TcxCustomGridRow* ARow = TableView->ViewData->Rows[i];
        if (ARow->Level == ChartView->ActiveDataLevel) {
          ARow->Focused = true;
          ARow->Selected = true;
        }
        else
          break;
      }
    }
  }  
  __finally {
    TableView->EndUpdate();
  }
}
//---------------------------------------------------------------------------

void TfrmMain::SynchronizeTableWithChart()
{
  UpdateTableGroupingAndColumnVisibility();
  ShowTableActiveGroup();
}
//---------------------------------------------------------------------------

void TfrmMain::UpdateTableGroupingAndColumnVisibility()
{
  TableView->BeginUpdate();
  try {
    TableView->Controller->ClearGrouping();
    for (int i = 0; i < ChartView->VisibleDataGroupCount; i++)
      GetColumnByChartItem(ChartView->VisibleDataGroups[i])->GroupIndex = i;

    for (int i = 0; i < ChartView->DataGroupCount; i++)
      GetColumnByChartItem(ChartView->DataGroups[i])->Visible = ChartView->DataGroups[i]->Visible;
    for (int i = 0; i < ChartView->SeriesCount; i++)
      GetColumnByChartItem(ChartView->Series[i])->Visible = ChartView->Series[i]->Visible;
  }    
  __finally {
    TableView->EndUpdate();
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::GridActiveTabChanged(TcxCustomGrid *Sender,
      TcxGridLevel *ALevel)
{
  if (ALevel == GridLevelTable)
    SynchronizeTableWithChart();
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::TableViewStylesGetGroupStyle(
      TcxGridTableView *Sender, TcxCustomGridRecord *ARecord, int ALevel,
      TcxStyle *&AStyle)
{
  if (ARecord == NULL) return;
  int ADataGroupIndex = TableView->DataController->Groups->DataGroupIndexByRowIndex[ARecord->Index];
  for (int i = 0; i < ActiveDataGroups.Length; i++)
    if (ADataGroupIndex == ActiveDataGroups[i]) {
      AStyle = styleActiveGroup;
      break;
    };
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/ChartDataDrillingDemo/ChartDataDrillingDemoMain.pas
unit ChartDataDrillingDemoMain;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, DB, Menus, StdCtrls, cxStyles, cxCustomData, cxGraphics,
  cxFilter, cxData, cxDataStorage, cxEdit, cxDBData, cxGridLevel, cxClasses,
  cxControls, cxGridCustomView, cxGridCustomTableView, cxGridTableView,
  cxGridDBTableView, cxGrid, cxGridChartView, cxGridDBChartView, cxTextEdit,
  cxLookAndFeels, cxLookAndFeelPainters, BaseForm, cxGridCardView, ComCtrls,
  DBClient, MidasLib;

type
  TfrmMain = class(TfmBaseForm)
    dsOrders: TDataSource;
    tblOrders: TClientDataSet;
    GridLevelChart: TcxGridLevel;
    Grid: TcxGrid;
    tblCustomers: TClientDataSet;
    tblProducts: TClientDataSet;
    tblOrdersID: TAutoIncField;
    tblOrdersCustomerID: TIntegerField;
    tblOrdersProductID: TIntegerField;
    tblOrdersPurchaseDate: TDateTimeField;
    tblOrdersTime: TDateTimeField;
    tblOrdersPaymentType: TStringField;
    tblOrdersPaymentAmount: TCurrencyField;
    tblOrdersDescription: TMemoField;
    tblOrdersQuantity: TIntegerField;
    tblOrdersCustomer: TStringField;
    tblOrdersProduct: TStringField;
    ChartView: TcxGridDBChartView;
    ChartViewSeriesPaymentAmount: TcxGridDBChartSeries;
    ChartViewDataGroupProduct: TcxGridDBChartDataGroup;
    GridLevelTable: TcxGridLevel;
    TableView: TcxGridDBTableView;
    ChartViewDataGroupPaymentType: TcxGridDBChartDataGroup;
    ChartViewDataGroupCustomer: TcxGridDBChartDataGroup;
    ChartViewDataGroupPurchaseDate: TcxGridDBChartDataGroup;
    ChartViewSeriesQuantity: TcxGridDBChartSeries;
    ChartViewSeriesCount: TcxGridDBChartSeries;
    TableViewPurchaseDate: TcxGridDBColumn;
    TableViewPaymentType: TcxGridDBColumn;
    TableViewPaymentAmount: TcxGridDBColumn;
    TableViewQuantity: TcxGridDBColumn;
    TableViewCustomer: TcxGridDBColumn;
    TableViewProduct: TcxGridDBColumn;
    TableViewOrderCount: TcxGridDBColumn;
    styleActiveGroup: TcxStyle;
    procedure GridActiveTabChanged(Sender: TcxCustomGrid; ALevel: TcxGridLevel);
    procedure TableViewStylesGetGroupStyle(Sender: TcxGridTableView;
      ARecord: TcxCustomGridRecord; ALevel: Integer; var AStyle: TcxStyle);
    procedure FormCreate(Sender: TObject);
  protected
    ActiveDataGroups: array of Integer;
    function GetColumnByChartItem(AChartItem: TcxGridChartItem): TcxGridDBColumn;
    procedure ShowTableActiveGroup;
    procedure SynchronizeTableWithChart;
    procedure UpdateTableGroupingAndColumnVisibility;
  end;

var
  frmMain: TfrmMain;

implementation

{$R *.dfm}

uses
  AboutDemoForm;

function TfrmMain.GetColumnByChartItem(AChartItem: TcxGridChartItem): TcxGridDBColumn;
begin
  Result := TableView.GetColumnByFieldName(TcxGridDBChartItemDataBinding(AChartItem.DataBinding).FieldName);
end;

procedure TfrmMain.ShowTableActiveGroup;
var
  I, ADataGroupIndex, AGroupRowIndex: Integer;
  ARow: TcxCustomGridRow;
begin
  // expand group rows so that the group with the active chart data is made visible
  SetLength(ActiveDataGroups, ChartView.ActiveDataLevel);
  TableView.BeginUpdate;
  try
    TableView.DataController.Groups.FullCollapse;
    ADataGroupIndex := -1;
    for I := 0 to ChartView.ActiveDataLevel - 1 do
    begin
      ADataGroupIndex := TableView.DataController.Groups.GetDataGroupIndexByGroupValue(
        ADataGroupIndex, ChartView.VisibleDataGroups[I].ActiveValue);
      ActiveDataGroups[I] := ADataGroupIndex;
        
      AGroupRowIndex := TableView.DataController.DataControllerInfo.DataGroups[ADataGroupIndex].RowIndex;
      TableView.DataController.Groups.ChangeExpanding(AGroupRowIndex, True, False);
      TableView.Controller.FocusedRowIndex := AGroupRowIndex;
    end;
  finally
    TableView.EndUpdate;
  end;
  // select rows with the data used for the active chart
  TableView.BeginUpdate;
  try
    if ADataGroupIndex = -1 then
      TableView.Controller.SelectAll
    else
    begin
      TableView.Controller.ClearSelection;
      for I := TableView.Controller.FocusedRowIndex + 1 to TableView.ViewData.RowCount - 1 do
      begin
        ARow := TableView.ViewData.Rows[I];
        if ARow.Level = ChartView.ActiveDataLevel then
        begin
          ARow.Focused := True;
          ARow.Selected := True;
        end
        else
          Break;
      end;
    end;
  finally
    TableView.EndUpdate;
  end;
end;

procedure TfrmMain.SynchronizeTableWithChart;
begin
  UpdateTableGroupingAndColumnVisibility;
  ShowTableActiveGroup;
end;

procedure TfrmMain.TableViewStylesGetGroupStyle(Sender: TcxGridTableView;
  ARecord: TcxCustomGridRecord; ALevel: Integer; var AStyle: TcxStyle);
var
  ADataGroupIndex, I: Integer;
begin
  if ARecord = nil then Exit;
  ADataGroupIndex := TableView.DataController.Groups.DataGroupIndexByRowIndex[ARecord.Index];
  for I := 0 to Length(ActiveDataGroups) - 1 do
    if ADataGroupIndex = ActiveDataGroups[I] then
    begin
      AStyle := styleActiveGroup;
      Break;
    end;
end;

procedure TfrmMain.UpdateTableGroupingAndColumnVisibility;
var
  I: Integer;
begin
  TableView.BeginUpdate;
  try
    TableView.Controller.ClearGrouping;
    for I := 0 to ChartView.VisibleDataGroupCount - 1 do
      GetColumnByChartItem(ChartView.VisibleDataGroups[I]).GroupIndex := I;

    for I := 0 to ChartView.DataGroupCount - 1 do
      GetColumnByChartItem(ChartView.DataGroups[I]).Visible := ChartView.DataGroups[I].Visible;
    for I := 0 to ChartView.SeriesCount - 1 do
      GetColumnByChartItem(ChartView.Series[I]).Visible := ChartView.Series[I].Visible;
  finally
    TableView.EndUpdate;
  end;
end;

procedure TfrmMain.GridActiveTabChanged(Sender: TcxCustomGrid; ALevel: TcxGridLevel);
begin
  if ALevel = GridLevelTable then
    SynchronizeTableWithChart;
end;

procedure TfrmMain.FormCreate(Sender: TObject);
var
  APath: string;
begin
  inherited;
  APath := ExtractFilePath(Application.ExeName) + '..\..\Data\';
  tblCustomers.LoadFromFile(APath + 'Customers.xml');
  tblProducts.LoadFromFile(APath + 'Products.xml');
  tblOrders.LoadFromFile(APath + 'Orders.xml');
end;

end.
